### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by evgenigourvitch**

### CODE REVIEW

File: uaparser/parser.go [hunk 1]
```diff
@@ -126,88 +192,124 @@ func NewFromBytes(data []byte) (*Parser, error) {
...
 	if !found {
 		dvc.Family = "Other"
 	}
+	if(foundIdx > matchIdxNotOk) {
+		parser.DeviceMisses++
+	}
 	return dvc
 }
 
+func checkAndSort(parser *Parser) {
```
Comment (grudzinski) [inline uaparser/parser.go:309]: Should be sync

Comment (evgenigourvitch) [inline uaparser/parser.go:309]: what?

File: uaparser/parser.go [hunk 2]
```diff
@@ -126,88 +194,130 @@ func NewFromBytes(data []byte) (*Parser, error) {
...
 	ua := new(UserAgent)
+	foundIdx := -1
 	found := false
-	for _, uaPattern := range parser.UA {
+	for i, uaPattern := range parser.UA {
 		uaPattern.Match(line, ua)
 		if len(ua.Family) > 0 {
 			found = true
+			foundIdx = i
+			atomic.AddUint64(&parser.UA[i].MatchesCount, 1)
```
Comment (grudzinski) [inline uaparser/parser.go]: atomic.AddUint64(&uaPattern.MatchesCount, 1)

File: uaparser/parser.go [hunk 3]
```diff
@@ -126,88 +194,130 @@ func NewFromBytes(data []byte) (*Parser, error) {
...
 	os := new(Os)
+	foundIdx := -1
 	found := false
-	for _, osPattern := range parser.OS {
+	for i, osPattern := range parser.OS {
 		osPattern.Match(line, os)
 		if len(os.Family) > 0 {
 			found = true
+			foundIdx = i
+			atomic.AddUint64(&parser.OS[i].MatchesCount, 1)
```
Comment (grudzinski) [inline uaparser/parser.go]: atomic.AddUint64(& osPattern.MatchesCount, 1)

File: uaparser/parser.go [hunk 4]
```diff
@@ -126,88 +194,130 @@ func NewFromBytes(data []byte) (*Parser, error) {
...
 	dvc := new(Device)
+	foundIdx := -1
 	found := false
-	for _, dvcPattern := range parser.Device {
+	for i, dvcPattern := range parser.Device {
 		dvcPattern.Match(line, dvc)
 		if len(dvc.Family) > 0 {
 			found = true
+			foundIdx = i
+			atomic.AddUint64(&parser.Device[i].MatchesCount, 1)
```
Comment (grudzinski) [inline uaparser/parser.go]: atomic.AddUint64(&dvcPattern.MatchesCount, 1)

File: uaparser/parser.go [hunk 5]
```diff
@@ -69,13 +84,19 @@ func (os *osParser) setDefaults() {
 	}
 }
 
+type DeviceSorter []*deviceParser
+func (a DeviceSorter) Len() int           { return len(a) }
+func (a DeviceSorter) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a DeviceSorter) Less(i, j int) bool { return a[i].MatchesCount > a[j].MatchesCount }
```
Comment (grudzinski) [inline uaparser/parser.go]: Should use atomic here

File: uaparser/parser.go [hunk 6]
```diff
@@ -40,6 +49,11 @@ func (ua *uaParser) setDefaults() {
 	}
 }
 
+type OsSorter []*osParser
+func (a OsSorter) Len() int           { return len(a) }
+func (a OsSorter) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a OsSorter) Less(i, j int) bool { return a[i].MatchesCount > a[j].MatchesCount }
```
Comment (grudzinski) [inline uaparser/parser.go]: Should use atomic here

File: uaparser/parser.go [hunk 7]
```diff
@@ -13,8 +15,14 @@ type RegexesDefinitions struct {
 	UA     []*uaParser     `yaml:"user_agent_parsers"`
 	OS     []*osParser     `yaml:"os_parsers"`
 	Device []*deviceParser `yaml:"device_parsers"`
+	sync.RWMutex
 }
 
+type UserAgentSorter []*uaParser
+func (a UserAgentSorter) Len() int           { return len(a) }
+func (a UserAgentSorter) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a UserAgentSorter) Less(i, j int) bool { return a[i].MatchesCount > a[j].MatchesCount }
```
Comment (grudzinski) [inline uaparser/parser.go]: Should use atomic here

File: uaparser/parser.go [hunk 8]
```diff
@@ -126,88 +194,130 @@ func NewFromBytes(data []byte) (*Parser, error) {
...
 	}
+	if(foundIdx > matchIdxNotOk) {
+		atomic.AddUint64(&parser.DeviceMisses, 1)
+	}
 	return dvc
 }
 
+func checkAndSort(parser *Parser) {
+	if(atomic.LoadUint64(&parser.UserAgentMisses) >= missesTreshold) {
+		parser.RLock()
```
Comment (grudzinski) [inline uaparser/parser.go]: Why RLock? You are going to change it now, you should Lock it. Read side should use RLock, but I don't see it in the code.

**Comment by orcaman** [comment]
@evgenigourvitch @aviramst can someone please close this PR? it's stuck in my assigned PRs on GitHub :-)

**Pull request closed without merging**